### PR TITLE
ci: improving readability of testing matrix

### PR DESCRIPTION
**Pull request by Borda**
As it is decalted now it is not trivial to make contributor head what are the tested cases...
So first sing a dictionary written in a single line to its closed-to-table reading which is easier for humans; then rather define rules for run with exceptions then special cases...

This also will increase the number of running suits, but I believe it is in favor of extension stability

cc: @bsipocz @pllim

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
Thanks for your contributon!

However, the updated matrix is much harder for me to read, which is opposite of "making things easier for contributors" as you claimed. But I am just one data point.

The `pytestoldest` naming and exact pinning was actually requested by @bsipocz .

The only thing that I 100% agree on right now is adding `pytest81: pytest==8.1.*` job.

File: .github/workflows/python-tests.yml [hunk 1]
```diff
@@ -87,4 +52,12 @@ jobs:
     - name: Install Tox
       run: python -m pip install tox
     - name: Run Tox
-      run: tox ${{ matrix.toxargs }} -v -e ${{ matrix.toxenv }}
+      env:
+        PY: ${{ matrix.python-version }}
+        PT: ${{ matrix.pytest-version }}
+      run: |
+        # crete the token as "pyXY-test-pytestXY" from the matrix with drop dots
```
Comment (pllim) [inline .github/workflows/python-tests.yml:59]: The comment has typo and I don't know what "drop dots" mean.

Comment (Borda) [reply .github/workflows/python-tests.yml:59]: the base python notation is `3.8` with dot but for env you use just numbers `38`

**Comment by pllim** [comment]
Thanks for the quick response! Let's see what @bsipocz say first. 🐱

File: .github/workflows/python-tests.yml (hunk 1 above)
Comment (bsipocz) [reply .github/workflows/python-tests.yml:59]: If I see correctly this matrix update will have way more jobs, as we haven't tested all old versions with all OSes, etc, and have already received criticism that there are way too many jobs for the versions in the middle.

File: tox.ini [hunk 2]
```diff
@@ -13,7 +13,7 @@ setenv =
     numpydev: PIP_EXTRA_INDEX_URL = https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
 description = run tests
 deps =
-    pytestoldest: pytest==4.6.0
+    pytest46: pytest==4.6.*
```
Comment (bsipocz) [inline tox.ini:16]: The oldest have to stay explicit and the version that is declared in the config. Besides that, there is no need to test the latest bugfix in the 4.6 series

**Comment by bsipocz** [comment]
I find the matrix part more legible, but it doesn't do what we had before. Otoh, I don't fund the curly brackets and longer lines etc around the one-off combinations helpful, quite the opposite.

**Review by bsipocz — CHANGES_REQUESTED** [review]
This has to be changed back. We explicitly use "oldest" in the name, but just here but a lot of other places to show it's the job with the oldest supported version of declared dependencies. 
The proposed change is something that let through bugs before we switched to be explicit.

**Comment by Borda** [comment]
@bsipocz not sure if I got your reply correctly, sounds like you do not agree with any part of this PR, correct? in such case, let's just close it... :)

**Comment by bsipocz** [comment]
I like the matrix form better but unfortunately it doesn't do the same thing, see a ci run of 19 jobs on main vs 157 jobs in this PR.

So if we take away those changes that don't work, unfortunately, it's indeed just the pytest addition that remains in this PR.

**Comment by Borda** [comment]
> I like the matrix form better but unfortunately it doesn't do the same thing, see a ci run of 19 jobs on main vs 157 jobs in this PR.

You are right as I did all combinations (and it was a demonstration not necessarily final state), but also you may want to define rich combinations for all Python and `pytest` versions on Ubuntu and then add a few includes for the latest and oldest configurations for MacOS and Windows.

**Comment by bsipocz** [comment]
Even that maybe an overkill. The only thing we definitely need to test for is all the (major) pytest versions as those sometimes change suddenly. Then the combination of the OS or python version can be much sparer as pytest itself already test for those. Given the incompatibilities (new pytest on old python vs old pytest on new python) the matrix form will get rather complicated to make it work, and then we'll need to add enough other OS jobs. 
So maybe this all comes down to some documentation lines in the config file that summarize the above logic, and also separates the "walking through the versions" jobs from the more special cases?

**Comment by bsipocz** [comment]
I didn't mean to close just yet it with the other PR.

**Pull request closed without merging**